### PR TITLE
Make sure "turn.locale" is correctly set in TurnContext, enabling locale can be changed during runtime

### PR DESCRIPTION
**Pull request by cosmicshuai**
Closes: #4983 #3993 
In this PR, fix a bug of set locale in TurnState. The locale info will be set to the value of key "turn".
User can set locale by SetProperty action.
For example, suppose the original locale is "en-US":
```
{
    "$schema": "../../tests.schema",
    "$kind": "Microsoft.AdaptiveDialog",
    "generator": "test.lg",
    "triggers": [
        {
            "$kind": "Microsoft.OnBeginDialog",
            "actions": [
                {
                    "$kind": "Microsoft.SetProperty",
                    "property": "turn.locale",
                    "value": "de-DE"
                },
                {
                    "$kind": "Microsoft.SendActivity",
                    "activity": "${string(1.122)}"
                },
                {
                    "$kind": "Microsoft.SendActivity",
                    "activity": "${formatNumber(1.12345, 4)}"
                },
                {
                    "$kind": "Microsoft.SendActivity",
                    "activity": "${addDays('2018-01-02T02:00:00.000Z', 4, 'D')}"
                },
                {
                    "$kind": "Microsoft.SendActivity",
                    "activity": "${formatNumber(3.1415926, 5)}"
                }
            ]
        }
    ]
}
```

### CODE REVIEW

**Review by stevengum — COMMENTED** [review]
This PR looks good to me, with the relevant context from @xieofxie: https://github.com/microsoft/botbuilder-dotnet/issues/3993#issuecomment-681633716 

I'm a little hesitant to say this fixes https://github.com/microsoft/botbuilder-dotnet/issues/4983 as the customer didn't indicate what exactly they were trying in their bot(s) to get and set the locale.

@mrivera-ms or @EricDahlvang can one of you take a look for .NET correctness?

File: tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Templates.Tests/Microsoft - Backup.Bot.Builder.Dialogs.Adaptive.Templates.Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,79 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (cosmicshuai) [reply tests/Microsoft.Bot.Builder.Dialogs.Adaptive.Templates.Tests/Microsoft - Backup.Bot.Builder.Dialogs.Adaptive.Templates.Tests.csproj]: This file is created unintentionally while I was editing csproj file. I deleted it.

**Comment by boydc2014** [comment]
> This PR looks good to me, with the relevant context from @xieofxie: [#3993 (comment)](https://github.com/microsoft/botbuilder-dotnet/issues/3993#issuecomment-681633716)
> 
> I'm a little hesitant to say this fixes #4983 as the customer didn't indicate what exactly they were trying in their bot(s) to get and set the locale.
> 
> @mrivera-ms or @EricDahlvang can one of you take a look for .NET correctness?

The customer's request (among other customers's requests like @xieofxie) is to provide an approach to change local in runtime.

We previously designed a way to do this, is to create a property in turn, turn.locale to be able to get\set, and to change locale during a turn. But there was some issue previously, causing this property is not writeable. This PR just fixed the bug.

**Comment by stevengum** [comment]
> > This PR looks good to me, with the relevant context from @xieofxie: [#3993 (comment)](https://github.com/microsoft/botbuilder-dotnet/issues/3993#issuecomment-681633716)
> > I'm a little hesitant to say this fixes #4983 as the customer didn't indicate what exactly they were trying in their bot(s) to get and set the locale.
> > @mrivera-ms or @EricDahlvang can one of you take a look for .NET correctness?
> 
> The customer's request (among other customers's requests like @xieofxie) is to provide an approach to change local in runtime.
> 
> We previously designed a way to do this, is to create a property in turn, turn.locale to be able to get\set, and to change locale during a turn. But there was some issue previously, causing this property is not writeable. This PR just fixed the bug.

This sounds reasonable to me!

File: libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs [hunk 2]
```diff
@@ -659,7 +659,7 @@ public async Task<bool> EmitEventAsync(string name, object value = null, bool bu
         /// <returns>A string representing the current locale.</returns>
         public string GetLocale()
         {
-            return Context.TurnState.Get<string>("turn.locale");
+            return (Context as TurnContext).Locale;
```
Comment (joshgummersall) [inline libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs:662]: This appears to be reverting a seemingly intentional change @carlosscastro made here: https://github.com/microsoft/botbuilder-dotnet/pull/4860

Comment (cosmicshuai) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs]: Right, I recalled this change in previous discussion. Due to the path is changed, I did some modification this time. But still follows Carlos' suggestion.

Comment (joshgummersall) [reply libraries/Microsoft.Bot.Builder.Dialogs/DialogContext.cs]: Good deal, just wanted to double check.

**Review by boydc2014 — APPROVED** [review]

**Review by mrivera-ms — APPROVED** [review]

**Pull request merged**